File: Client-Java/src/main/java/pt/ulisboa/tecnico/tuplespaces/client/ClientMain.java
package pt.ulisboa.tecnico.tuplespaces.client;

import pt.ulisboa.tecnico.tuplespaces.client.grpc.ClientService;

public class ClientMain {

    private static final String SEPARATOR = ":";

    public static void main(String[] args) {

        System.out.println(ClientMain.class.getSimpleName());

        // checks to see if -debug flag is present (last position assumed)
        boolean debugFlag = false;
        int numArgs = args.length;

        // check arguments
        if (numArgs < 2 || numArgs > 3 || (numArgs == 3 && !args[numArgs - 1].equals("-debug"))) {
            System.err.println("Invalid number of arguments!");
            System.err.println("Usage: mvn exec:java -Dexec.args=<host:port> <client_id> [-debug]");
            return;
        }

        if (args[numArgs - 1].equals("-debug")) {
            debugFlag = true;
        }

        // receive and print arguments
        if (debugFlag) {
            System.out.printf("Received %d arguments%n", args.length);
            for (int i = 0; i < args.length; i++) {
                System.out.printf("arg[%d] = %s%n", i, args[i]);
            }
        }

        // get the host and the port
        final String hostPort = args[0];

        String[] parts = hostPort.split(SEPARATOR);

        // Ensure the format is correct (host:port)
        if (parts.length != 2) {
            System.err.println("Error: Invalid format for host:port -> " + hostPort);
            return;
        }

        // Extract and validate the port and client ID
        final int port;
        final int clientId;
        try {
            port = Integer.parseInt(parts[1]);
            clientId = Integer.parseInt(args[1]);
        } catch (NumberFormatException e) {
            System.err.println("Error: Port and Client ID must both be integers!");
            return;
        }

        // Create the client service
        ClientService clientService = new ClientService(hostPort, clientId, debugFlag); 

        CommandProcessor parser = new CommandProcessor(clientService);
        parser.parseInput();

        // Shutdown the client properly
        clientService.shutdown();
    }
}

File: Client-Python/client_main.py
import sys
from typing import List
from client_service import ClientService
from command_processor import CommandProcessor

class ClientMain:
    SEPARATOR = ":"

    @staticmethod
    def main(args: List[str]):
        print("ClientMain")

        # checks to see if -debug flag is present (last position assumed)
        debug_flag = False
        num_args = len(args)

        # check arguments
        if num_args < 2 or num_args > 3 or (num_args == 3 and not (args[num_args - 1] == "-debug")) :
            print("Invalid number of arguments!")
            print("Usage: python3 client_main.py <host:port> <client_id> [-debug]")
            return
        
        if args[num_args - 1] == "-debug":
            debug_flag = True
            # receive and print arguments
            print(f"Received {len(args)} arguments")
            for i, arg in enumerate(args):
                print(f"arg[{i}] = {arg}")

        # Get the host and port of the server or front-end
        host_port = args[0]
        parts = host_port.split(ClientMain.SEPARATOR)

        if len(parts) != 2:
            print(f"Error: Invalid format for host:port -> {host_port}", file=sys.stderr)
            return
        try:
            port = int(parts[1])
            client_id = int(args[1])
        except ValueError:
            print("Error: Port and Client ID must both be integers!", file=sys.stderr)
            return        

        client_service = ClientService(host_port, client_id, debug_flag)

        parser = CommandProcessor(client_service)
        parser.parse_input()
        
        client_service.close()


if __name__ == "__main__":
    ClientMain.main(sys.argv[1:])